lib/web/ui/server.go

```diff
@@ -88,19 +93,55 @@ func MakeServers(clusterName string, servers []types.Server) []Server {
 
 		sort.Sort(sortedLabels(uiLabels))
 
+		sshLogins := getServerLogins(server, userRoles)
+
 		uiServers = append(uiServers, Server{
 			ClusterName: clusterName,
 			Labels:      uiLabels,
 			Name:        server.GetName(),
 			Hostname:    server.GetHostname(),
 			Addr:        server.GetAddr(),
 			Tunnel:      server.GetUseTunnel(),
+			SSHLogins:   sshLogins,
 		})
 	}
 
 	return uiServers
 }
 
+// getServerLogins returns the list of logins the role set has that is associated with this node
+func getServerLogins(server types.Server, roleSet services.RoleSet) []string {
+	allowed := []string{}
+	denied := []string{}
+	for _, role := range roleSet {
+		// remove any denied logins, even if they don't apply to that node's labels
+		denied = append(denied, role.GetLogins(types.Deny)...)
+
+		// add only logins related to that server
+		isAllowed, _, err := services.MatchLabels(role.GetNodeLabels(types.Allow), server.GetAllLabels())
```

Comment:
This is reimplementing logic from `lib/services/role.go`. When working on related PR: https://github.com/gravitational/teleport/pull/10458, I made similar change and received feedback that these should go to the core of Teleport or be exposed via endpoint.

I went with the former and added changes here: https://github.com/gravitational/teleport/pull/10458/files#diff-9b3fca74b2b2d4ce46a466de333f6c3e7640c21da8bd587e1837aa7a113eb7e7

I think you can reimplement `getServerLogins` in a similar fashion to how `EnumerateDatabaseUsers` works, unless there are issues I cannot see now.

Reply:
Thanks, that makes sense.

I moved the method to `lib/services/role.go` and implemented it like `EnumerateDatabaseUsers` on https://github.com/gravitational/teleport/pull/13266/commits/da94e407094841fdef1ca2902ae3902d9e639038.

I didn't find a good way to reuse the logic of `EnumerateDatabaseUsers`, even though the methods are mostly the same, so I ended up duplicating it.

We could abstract things like `role.GetDatabaseUsers / role.GetLogins` to a single method like `GetResourceUsers(resource string, type types.RoleConditionType)` but IDK if that would make sense, DB users and node logins aren't precisely the same thing.

I think it's better to duplicate the logic here than to make something overly complicated.
Please let me know what you think.

Reply:
Thank you, this shapes up to be something I'd expect to see. 

I don't think the duplication is too bad as it is. For better or worse, we have a lot of duplication in that part of code already 🤷 . While repetitive in some cases, it gives us an easy way to be flexible in others: case in point with the wildcards treatment.